core/src/main/java/org/jruby/truffle/runtime/core/RubyArray.java

```diff
@@ -244,8 +216,11 @@ public int getSize() {
     }
 
     public void setSize(int size) {
+        assert verifyStore(this.store, size);
```

Comment:
There are the following effects:
1) The interpreter will execute less # of bytecodes. The check of the assert boolean is not stripped by javac. So the interpreter will just go ahead and check the assert boolean all over.
2) The inlining heuristic of C1 relies on # bytecodes in a method. C1 will think this is a huge method while large parts of it immediately disappear when the asserts are removed by canonicalization.
3) C1 and also Graal will first parse the bytecodes, convert them into their IR, and then see that large parts of the IR is dead code. This means unnecessarily long compilation times.

The effect is not really large, but it adds up. For the asserts, one option is to strip them with a static tool. For general slow path code, this is not possible to do automatically.